### PR TITLE
NAS-107845 / 20.10 / make device.get_disks better for SCALE HA

### PR DESCRIPTION
**Pull request by yocalebo**
On SCALE HA, I've noticed the following:

1. on a functional HA system, the webUI was taking 5+ seconds to present the login screen
2. the standby nodes console was being spammed with `reservation conflict` messages

I traced this all down to the `device.get_disks` method. This method was calling `lshw` which was taking 2+ seconds to run on a system with 53 SCSI disks and causing the `reservation conflict` messages. I haven't looked at that source code but it seemed to be "loading" all subsystems no matter if you chose the "DISK" type. It is also opening the underlying disks in "w" mode which is causing the reservation errors. Instead of running `lshw`, I've replaced it with `lsblk`. The only major difference is that I'm having to shell out and run `sg_vpd` to grab the rotational rate of the disk if the type is `HDD`. Even doing this, my local system is ~52% faster and if type is `SSD`, then it can be ~85%+ faster since we don't have to grab the `rotationrate` attribute.

I've removed the private `retrieve_lshw_disks_data` method as there are no public consumers of that method.

`device.get_disks` before the change = 2.3 seconds
`device.get_disks` after the change = 1.1 seconds

`failover.disabled_reasons` before the change = 2.5 seconds
`failover.disabled_reasons` after the change = 1.1 seconds

### CODE REVIEW

**Comment by yocalebo** [comment]
I should add that we could make this exponentially faster by caching the disk information on middlewared setup, and then listening for uevent changes for add/remove on the disks and updating/invalidating the cache as necessary depending on the uevent. That would be the "proper" fix but this was a relatively easy stop-gap for now.

**Review by william-gr — APPROVED** [review]
We can probably wait to merge this after 20.10-ALPHA?

**Comment by yocalebo** [comment]
> We can probably wait to merge this after 20.10-ALPHA?

yes, we can wait 👍

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/device_/device_info_linux.py:189]: I think we should still set `mediasize` for backwards compatibility on consumers consuming that key. We do that for freebsd side as well.

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/device_/device_info_linux.py:178]: This block is not required as `lsblk` retrieves the model data automatically for nvme drives

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/device_/device_info_linux.py:194]: I am not sure about this. `lunid` for a nvme drive i have right now on my scale system is
```
0025385201416c20
```
`sg_vpd` gives
```
sg_vpd -i --quiet /dev/nvme0n1
  0x0025385201416c20
  eui.0025385201416C20
```

With current implementation we take the first id available. Now `lsblk` reports the second id shown in `sg_vpd` as `wwn`.
What I think will happen is that middleware will probably recognise the nvme as a new disk because it won't be able to translate `serial_lunid` to nvme disk name as the id would have changed for the nvme drive.

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/device_/device_info_linux.py:131]: How about using regex or checking key name like `Nominal rotation rate` rather then relying on `1` index as it could probably change ? We are already checking `rotation rate` in if below but we might lose it here before as it might on a different index

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py:194]: I think removing `eui.` and having it lower case should be enough ( though we should confirm if it won't change in any other way )

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (yocalebo) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py:189]: oops, missed a key. good catch, I'll fix.

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (yocalebo) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py:131]: sg_vpd has a very stable API. If this was the case, then we would also need to do a regex for the lunid section. I'd rather leave this as is.

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py:131]: Why the if check below then ? ; )
My concern was that we do a check partially, so it shows we are not sure of the output, but doing the check there means we already have potentially lost the correct line.

Comment (yocalebo) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py:131]: Because `Nominal rotation rate:` doesn't get returned on SSD's.

Comment (yocalebo) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py:131]: Even though we're only doing this on HDD types, having that extra `if` is to make sure that `lsblk` identified the type correctly and passed us a HDD.

Comment (yocalebo) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py]: I don't like it, but to get your approval I've changed it to use a REGEX 👍

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/device_/device_info_linux.py]: Thank you for clarifying and using regex @yocalebo : ) I was okay with it if you think it's good :p

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/device_/device_info_linux.py:196]: Can we please just have this and remove the if block at L183 in the new diff ? As if we get into last `if` block, we will miss setting `serial_lunid`.

File: src/middlewared/middlewared/plugins/device_/device_info_linux.py
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/device_/device_info_linux.py:179]: We have the prefix `eui.` which we should be removing. Looking at `lsblk` output again, i don't think we need lower, `sg_vpd` identification flag returns them capitalised, lsblk seems to be doing the right thing following what we have been doing before.

**Review by sonicaj — APPROVED** [review]
Thank you

File: src/freenas/usr/local/lib/middlewared_truenas/plugins/failover.py
Comment (sonicaj) [inline src/freenas/usr/local/lib/middlewared_truenas/plugins/failover.py:553]: It just came to me, `retrieve_disks_data ` this does not exist in core device service, are we looking for freebsd support in failover code for master ?

Comment (yocalebo) [reply src/freenas/usr/local/lib/middlewared_truenas/plugins/failover.py:553]: You're correct, I don't like calling `device.get_info 'DISK'` because it gets the rotationrate of every drive (by subprocessing out) which causes slowness. However, I don't see an easy way around it without making public API changes or doing an `if osc.IS_LINUX` check which is gross. I'll just go back to using the `get_info 'DISK'` API for now.

**Pull request merged**
